src/klein/app.py

```diff
@@ -186,9 +194,37 @@ def index(request):
 
         @returns: decorated handler function.
         """
-        segment_count = url.count('/')
-        if url.endswith('/'):
-            segment_count -= 1
+
+        def _urlToString(_url):
+            return _url.asText() if URL is not False else _url
+
+        if URL is not False:
+
+            if isinstance(url, string_types + (binary_type, )):
+                url = URL.fromText(url)
+            if url.absolute or url.port or url.userinfo:
+                warnings.warn(
+                    "Mind to use base URI for an application routing, "
+                    "instead of "
+                    + url.asText(includeSecrets=False))
```

Comment:
How about raising `ValueError` here instead?  I'm not a fan of warnings for what is probably a bug in the caller, and less so of cleaning up below to do what we think maybe the caller wants.